Introduce v1alpha2 kube proxy configuration

aroradaman opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR introduces `v1alpha2` configuration for kube-proxy.
https://kep.k8s.io/784 for more details. 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #117909

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/784-kube-proxy-component-config
```

aroradaman commented:
/test pull-kubernetes-e2e-kind

aroradaman commented:
/test pull-kubernetes-e2e-gce

aroradaman commented:
/milestone v1.30

aroradaman commented:
@danwinship @aojea tested out with [this commit on top of kind main](https://github.com/aroradaman/kind/commit/df1e2f2da18284f0a48440258742c7ebbe2111e8), was able to provision a local kind cluster with `kubeproxy.config.k8s.io/v1alpha2`

aroradaman commented on cmd/kubeadm/app/componentconfigs/kubeproxy.go:
```diff
@@ -20,7 +20,7 @@ import (
 	"github.com/pkg/errors"
 	clientset "k8s.io/client-go/kubernetes"
 	"k8s.io/klog/v2"
-	kubeproxyconfig "k8s.io/kube-proxy/config/v1alpha1"
```
I'm not sure how to make kubeadm work with both v1alpha1 and v1alpha2 kube proxy config

neolit123 replied:
this change needs a discussion in a KEP.

the original linked KEP never answered the questions on how to migrate users from one version to another.
if we know the answers we can draft a plan.

right now this is just a PR that introduces some good features, but can break and/or confuse many users.

neolit123 replied:
a summary for kubeadm to support this:
- on kubeadm init  / cluster creation v1alpha2 must be used and stored in the config map
- kubeadm defaults must be patched everywhere to use the new API version
- on kubeadm upgrade ,the kube-proxy config map stored in the cluster must be mutated to v1alpha2

proposal:

ideally kube-proxy must export a CLI option such as `kube-proxy --convert-config-input --convert-config-output` that can take a v1alpha1 and produce a v1alpha2, this is ALA `kubectl convert`.
only kube-proxy (the component binary) keeps track of it's internal config and can perform this conversion.

the command can accept a set of API structs separated by `---` but only migrated the known kube-proxy API kinds.

this process must be documented somewhere at the k8s.io website.
i can comment this again on the KEP update PR, once it's open.

neolit123 replied:
that's great. good to know.
it just has to be documented at the right places at the k/website.

aroradaman replied:
https://github.com/kubernetes/enhancements/pull/4337

danwinship commented on pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha2.yaml:
```diff
@@ -0,0 +1,61 @@
+detectLocal:
+  bridgeInterface: ""
+  clusterCIDR: ""
+  interfaceNamePrefix: ""
```
We really ought to make better use of "omitempty" in v1alpha2.

danwinship commented on pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha2.yaml:
```diff
@@ -0,0 +1,61 @@
+enableProfiling: false
+healthzBindAddress: 0.0.0.0:10256
+hostnameOverride: ""
+iptables:
+  localhostNodePorts: true
```
So one of the ideas was that we were going to change the defaults in v1alpha2, eg to make `localhostNodePorts` `false`. (But this is one of the things that would be discussed in the KEP update that, as @neolit123 mentioned, we still need to do.)

aroradaman commented on pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/roundtrip/default/v1alpha1.yaml:
```diff
@@ -32,7 +32,7 @@ iptables:
 ipvs:
   excludeCIDRs: null
-  minSyncPeriod: 0s
+  minSyncPeriod: 1s
```
We do not set any default for IPVS.MinSyncPeriod https://github.com/kubernetes/kubernetes/blob/5031ae11590748f247c15c81f9f35390a786104b/pkg/proxy/apis/config/v1alpha1/defaults.go#L62-L79

If I don't make this change then the round trip test fails as we consolidate all `minSyncPeriod` in the internal config and then push to the `minSyncPeriod` field of the respective backend in v1alpha1 config.

danwinship commented on pkg/proxy/apis/config/types.go:
```diff
@@ -184,22 +180,25 @@ type KubeProxyConfiguration struct {
+	// healthzBindAddresses is a list of CIDR ranges that contains a valid node IP on which
+	// health check server will be served on, defaulting to [ "0.0.0.0/0", "::/0" ].
+	HealthzBindAddresses []string
+	// healthzBindPort is the port on which helathz server will be exposed, defaulting to 10256.
```
typo "healthz"

danwinship commented on pkg/proxy/apis/config/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,159 @@
+	} else {
+		out.DetectLocal.ClusterCIDRs = []string{}
+	}
+	out.NodeIPOverride = strings.Split(in.BindAddress, ",")
```
`v1alpha1.BindAddress` didn't allow comma-separated values.

also, for historical reasons, `BindAddress` always had to be set, and defaulted to `0.0.0.0`, but `NodeIPOverride` sounds more like it should just be unset/empty if you aren't overriding the node IP

danwinship commented on pkg/proxy/apis/config/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,159 @@
+	out.NFTables.MasqueradeAll = in.Linux.MasqueradeAll
+	out.IPTables.MasqueradeAll = in.Linux.MasqueradeAll
+
+	ip, _, _ := netutils.ParseCIDRSloppy(in.HealthzBindAddresses[0])
```
is `in.HealthzBindAddress` guaranteed to be non-empty after defaulting?

danwinship commented on pkg/proxy/apis/config/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,159 @@
+	ip, _, _ = netutils.ParseCIDRSloppy(in.MetricsBindAddresses[0])
+	out.MetricsBindAddress = net.JoinHostPort(ip.String(), strconv.Itoa(int(in.MetricsBindPort)))
+
+	out.BindAddress = in.NodeIPOverride[0]
```
`NodeIPOverride` is definitely not guaranteed to be non-empty

aroradaman replied:
we do have default values for bind, metrics and healthz address in v1alpha1 and will have the same defaults in v1alpha2 as we are just changing the data type from string to list[string].

Do we need to handle empty string/list cases if we know that the underlying value will always be non-empty?

https://github.com/kubernetes/kubernetes/blob/12217672a3511f5e461ecb87b3e22422ab30bdbf/pkg/proxy/apis/config/v1alpha1/defaults.go#L41-L56

danwinship replied:
> we do have default values for bind, metrics and healthz address in v1alpha1 and will have the same defaults in v1alpha2 as we are just changing the data type from string to list[string].

That doesn't follow. Defaulting is per-version. You can have `v1alpha1` default `BindAddress` to `"0.0.0.0"` but have `v1alpha2` not have any default for `NodeIPOverride`.

aroradaman replied:
`HealthzBindAddresses` and `MetricsBindAddresses` will default to [0.0.0.0/0, ::/0] and [127.0.0.0/8, ::1/128] in v1alpha2, right?

And `HealthzBindPort` and `MetricsBindPort` will be 10256 and 10249 same as before?

aroradaman replied:
Can the defaulting be done conditionally based on the IPFamily of the cluster?
We do check IP/CIDR configuration while initializing the proxy, and with `ConfigHardFail` default to true that will fail for v1alpha2 if we default healthz/metrics addresses with dual-stack values and the cluster is not dual-stack.

danwinship replied:
> Can the defaulting be done conditionally based on the IPFamily of the cluster?

not really... defaulting can't look at anything outside the object itself, and the object doesn't (currently) unambiguously indicate the IP family.

you could potentially add a flag for that, though it would have to be optional since there's no v1alpha1 equivalent

danwinship commented on pkg/proxy/apis/config/validation/validation.go:
```diff
@@ -54,128 +53,78 @@ func Validate(config *kubeproxyconfig.KubeProxyConfiguration) field.ErrorList {
+		if netutils.ParseIPSloppy(addr) == nil {
+			allErrs = append(allErrs, field.Invalid(newPath.Child("NodeIPOverride").Index(i), addr, "not a valid textual representation of an IP address"))
 		}
 	}
```
if it's not empty, it has to be either a single IP, or a dual-stack pair

danwinship commented on pkg/proxy/apis/config/validation/validation.go:
```diff
@@ -54,128 +53,78 @@ func Validate(config *kubeproxyconfig.KubeProxyConfiguration) field.ErrorList {
+	if len(config.HealthzBindAddresses) > 0 {
+		allErrs = append(allErrs, validateDualStackCIDRStrings(config.HealthzBindAddresses, newPath.Child("HealthzBindAddresses"))...)
+	}
+	if config.HealthzBindPort != 0 {
```
validation happens after defaulting, so it won't be 0

danwinship commented on pkg/proxy/apis/config/validation/validation.go:
```diff
@@ -351,3 +300,55 @@ func validateInterface(iface string, fldPath *field.Path) field.ErrorList {
+	case len(cidrs) == 0:
+		allErrs = append(allErrs, field.Invalid(fldPath, cidrs, "must contain at least one CIDR"))
+	case len(cidrs) > 2:
+		allErrs = append(allErrs, field.Invalid(fldPath, cidrs, "must be a valid CIDR list (e.g. [10.100.0.0/16, fde4:8dba:82e1::/48]"))
```
a list of more than 2 CIDRs is a "valid CIDR list", it's just not allowed in this context. you need to say "must be either a single CIDR or a dual-stack pair of CIDRs" or somethign like that

danwinship commented on pkg/proxy/apis/config/validation/validation.go:
```diff
@@ -351,3 +300,55 @@ func validateInterface(iface string, fldPath *field.Path) field.ErrorList {
+	case kubeproxyconfig.LocalModeInterfaceNamePrefix:
+		allErrs = append(allErrs, validateInterface(config.InterfaceNamePrefix, fldPath.Child("InterfacePrefix"))...)
+	case kubeproxyconfig.LocalModeClusterCIDR:
+		allErrs = append(allErrs, validateDualStackCIDRStrings(config.ClusterCIDRs, fldPath.Child("ClusterCIDRs"))...)
```
This is the only place I noticed it, but make sure you aren't accidentally tightening up v1alpha1 validation.

In particular, if `ConfigHardFail` isn't set, then you need to allow setting `LocalModeClusterCIDR` even if `ClusterCIDRs` is unset.

danwinship reviewed (commented):
yes, looks like a good start

aroradaman commented on pkg/proxy/apis/config/fuzzer/fuzzer.go:
```diff
@@ -33,19 +33,22 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
+			obj.Linux.Conntrack.TCPEstablishedTimeout = &metav1.Duration{Duration: time.Duration(c.Int63()) * time.Hour}
 			obj.FeatureGates = map[string]bool{c.RandString(): true}
-			obj.HealthzBindAddress = fmt.Sprintf("%d.%d.%d.%d:%d", c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(65536))
+			obj.HealthzBindAddresses = []string{fmt.Sprintf("%d.%d.%d.%d/32", c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(256))}
```
@danwinship should we run two instances of healthz/metrics server if the user provides a valid dual stack CIDR list and the node has local IPs for both the IP families?

danwinship replied:
Two healthz servers would undo your previous step toward getting dual-stack health working. You should extend `ProxierHealthServer` to support multiple bind addresses the same way `ServiceHealthServer` does.

Likewise with the metrics server in `serveMetrics`.

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -159,9 +171,14 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
+	fs.StringArrayVar(&o.config.NodeIPOverride, "node-ip-override", o.config.NodeIPOverride, "Overrides kube-proxy's idea of what its node's primary IP is. This parameter is ignored if a config file is specified by --config.")
+	fs.Var(&utilflag.IPVar{Val: &o.bindAddress}, "bind-address", "Overrides kube-proxy's idea of what its node's primary IP is. Note that the name is a historical artifact, and kube-proxy does not actually bind any sockets to this IP. This parameter is ignored if a config file is specified by --config.")
+	fs.Var(&utilflag.IPPortVar{Val: &o.healthzAddress}, "healthz-bind-address", "The IP address and port for the health check server to serve on, defaulting to \"0.0.0.0:10256\" (if --bind-address is unset or IPv4), or \"[::]:10256\" (if --bind-address is IPv6). Set empty to disable. This parameter is ignored if a config file is specified by --config.")
+	fs.StringArrayVar(&o.config.HealthzBindAddresses, "healthz-bind-addresses", o.config.HealthzBindAddresses, "The list is a list of CIDR ranges that contains a valid node IP on which health check server will be served on. Set empty to disable. This parameter is ignored if a config file is specified by --config.")
```
Hm... I had thought we discussed this but I guess it didn't make it into the KEP update...

We absolutely do not want "This parameter is ignored if a config file is specified" for v1alpha2. That's really really confusing, and we would have fixed it, except that it would be a config API break at this point. But v1alpha2 gives us the opportunity to do that.

So we should do one of:
1. If you use a `v1alpha2` config file, then you can't pass any other command-line flags (ie, kube-proxy will exit with an error). Thus, there is no need to add new command-line flags for the new v1alpha2 options.
2. Like 1, except a few command-line flags are allowed for inherently-per-node config options (in particular, `--node-ip-override` and `--hostname-override`. Maybe nothing else.) In that case the command-line flags override the config file, not vice versa.
3. Like 2, except _all_ flags are allowed, and there are new flags for new options, and all command-line flags override a `v1alpha2` config file (but _not_ a `v1alpha1` config file).

I think I was assuming 2. Not sure what other people think. (@aojea?)

FWIW there are no command-line flags for nftables config, and nobody has complained (or even _commented_ on that fact). I think very few people deploy kube-proxy "by hand", and all of the k8s deploy/install tools (kubeadm, etc) write out config files because it's cleaner...

aroradaman replied:
I like the second option, kubelet is also doing the same thing - https://github.com/kubernetes/enhancements/pull/2158#discussion_r616788544
https://github.com/kubernetes/kubernetes/blob/24267f6aa78d0b5b957dc618c7c4dac5dfa21fa3/cmd/kubelet/app/options/options.go#L335-L363

aroradaman replied:
Should we then also mark all the existing flags except logging/instance specific deprecated?

danwinship replied:
we can't do that yet; kubelet has a v1 config format

aroradaman replied:
understood!
If we don't deprecate the existing flags yet and don't add any new flags then we are good, right?
(and only allow select flags [node-ip-override, host-name-override] more precedence over config)

danwinship replied:
> FWIW there are no command-line flags for nftables config, and nobody has complained (or even _commented_ on that fact).

Antonio eventually complained/commented; [the GCE setup scripts](https://github.com/kubernetes/kubernetes/blob/37ca0377d542d9eb5b6a7ebac2093b8ec6fc9942/cluster/gce/gci/configure-helper.sh#L1781) (which are used for k8s CI among other things) use command-line flags rather than config files. (Because they predate the addition of `--config`...)

thockin replied:
I think option 2 makes sense, and I think it is important to behave similarly to other components (kubelet and kube-proxy are like cousins).  To recap, tell me if I understood correctly:

* Most options would be configurable EITHER by flags or a config file, but not both (applying to the whole set, not individually)
* A small set of options would be configurable by flags AND a config file (who wins if both are specified?)
* The only flag which is not also a config file option is `--config`

bart0sh commented:
/remove-sig node

leilajal commented:
/remove-sig api-machinery

aroradaman commented:
/remove-sig apps
/remove-sig architecture
/remove-sig auth
/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig instrumentation
/remove-sig network
/remove-sig release
/remove-sig scalability
/remove-sig scheduling
/remove-sig storage
/remove-sig testing

aroradaman commented:
/sig network

aroradaman commented:
/retest

danwinship commented on cmd/kube-proxy/app/options_test.go:
```diff
@@ -180,81 +180,83 @@ nodePortAddresses:
-			t.Fatalf("unexpected config for %s, diff = %s", tc.name, diff)
-		}
+			if diff := cmp.Diff(config, expected); diff != "" {
+				t.Fatalf("unexpected config for %s, diff = %s", tc.name, diff)
```
When `t.Run`-ifying a test case, you should remove the names (or numbers) from the existing log messages since they're redundant now. So this can be just `"unexpected config, diff = %s"`. (There may be other log/error messages to update too.)

danwinship commented on cmd/kube-proxy/app/options.go:
```diff
@@ -121,13 +133,13 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 
 	fs.Int32Var(o.config.IPTables.MasqueradeBit, "iptables-masquerade-bit", ptr.Deref(o.config.IPTables.MasqueradeBit, 14), "If using the iptables or ipvs proxy mode, the bit of the fwmark space to mark packets requiring SNAT with.  Must be within the range [0, 31].")
-	fs.BoolVar(&o.config.IPTables.MasqueradeAll, "masquerade-all", o.config.IPTables.MasqueradeAll, "If using the iptables or ipvs proxy mode, SNAT all traffic sent via Service cluster IPs. This may be required with some CNI plugins.")
+	fs.BoolVar(&o.config.Linux.MasqueradeAll, "masquerade-all", o.config.Linux.MasqueradeAll, "If using the linux proxy mode, SNAT all traffic sent via Service cluster IPs. This may be required with some CNI plugins.")
```
"linux" isn't a "proxy mode". I'd probably say: `"SNAT all traffic sent via Service cluster IPs. This may be required with some CNI plugins. Only supported on Linux."`

danwinship commented on cmd/kube-proxy/app/options.go:
```diff
@@ -138,36 +150,28 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
-	fs.Int32Var(&o.metricsPort, "metrics-port", o.metricsPort, "The port to bind the metrics server. Use 0 to disable.")
-	_ = fs.MarkDeprecated("metrics-port", "This flag is deprecated and will be removed in a future release. Please use --metrics-bind-address instead.")
-	fs.Var(utilflag.PortRangeVar{Val: &o.config.PortRange}, "proxy-port-range", "This was previously used to configure the userspace proxy, but is now unused.")
-	_ = fs.MarkDeprecated("proxy-port-range", "This flag has no effect and will be removed in a future release.")
```
(Make sure that the fact that these are finally being removed is documented in the release note, etc.)

danwinship commented on cmd/kube-proxy/app/options.go:
```diff
@@ -292,9 +294,9 @@ func (o *Options) errorHandler(err error) {
-	// Check if hostname-override flag is set and use value since configFile always overrides
+// processMaintainedFlags checks if maintained flags are set and, if so, updates the config with
+// the specified value giving precedence to these flags over config file.
+func (o *Options) processMaintainedFlags() error {
```
I'm not sure why you are calling these "maintained". I think that may not be the word you're looking for...

aroradaman replied:
I was struggling with the name, this is to process flags which will always have precedence over config, unlike the other flags. Will come up with a better name.

neolit123 replied:
if for v1alpha2 all flags override config, great.
kubelet and kubeadm do that. kube-proxy was with only some flags overriding.

danwinship commented on cmd/kube-proxy/app/options.go:
```diff
@@ -303,9 +305,56 @@ func (o *Options) processHostnameOverrideFlag() error {
 
+// processIncompatibleFlags processes flags which can't be directly mapped to internal config.
+func (o *Options) processIncompatibleFlags(fs *pflag.FlagSet) {
+	fs.Visit(func(f *pflag.Flag) {
```
This is weird...

Is this so that you can figure out which `Options` fields actually got set vs which ones just have the default values? Is there no better way to do that?

aroradaman replied:
Yes. Will try to find a better approach.

aroradaman replied:
have used https://pkg.go.dev/github.com/spf13/pflag@v1.0.5#FlagSet.Changed now

danwinship commented on cmd/kube-proxy/app/options.go:
```diff
@@ -214,17 +211,22 @@ func (o *Options) Complete(fs *pflag.FlagSet) error {
-		copyLogsFromFlags(fs, &c.Logging)
+		if err = copyLogsFromFlags(fs, &c.Logging); err != nil {
+			return err
+		}
```
(a separate commit to fix the lint problems caused by moving the code wouldn't be terrible...)

danwinship commented on cmd/kube-proxy/app/options_test.go:
```diff
@@ -379,6 +432,110 @@ func TestProcessHostnameOverrideFlag(t *testing.T) {
 }
 
+// TestProcessMaintainedFlags tests processing maintained flags.
+func TestProcessIncompatibleFlags(t *testing.T) {
```
wrong function name in comment

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -248,7 +248,7 @@ func newProxyServer(ctx context.Context, config *kubeproxyconfig.KubeProxyConfig
 	err, fatal := checkBadIPConfig(s, dualStackSupported)
 	if err != nil {
-		if fatal {
+		if fatal || (config.ConfigHardFail != nil && *config.ConfigHardFail) {
```
you should update the defaulting (`pkg/proxy/apis/config/*/defaults.go`) to fill this in if it's not set so that you can just assume it's not nil here

aroradaman replied:
v1alpha1 won't have `ConfigHardFail`. So if a user comes with v1alpha1 then then config.ConfigHardFail will be always nil, right?
(config being internal config after conversion)

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -341,14 +339,14 @@ func checkBadIPConfig(s *ProxyServer, dualStackSupported bool) (err error, fatal
 	if !dualStackSupported {
 		if badCIDRs(s.Config.IPVS.ExcludeCIDRs, badFamily) {
-			errors = append(errors, fmt.Errorf("cluster is %s but ipvs.excludeCIDRs contains only IPv%s addresses", clusterType, badFamily))
+			errors = append(errors, fmt.Errorf("cluster is %s but healthzBindAddresses contains only IPv%s addresses", clusterType, badFamily))
```
Bad copy/paste? I think you meant to replace the error message below but you replaced this one instead

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -464,20 +438,36 @@ func serveMetrics(bindAddress string, proxyMode kubeproxyconfig.ProxyMode, enabl
-		err := http.ListenAndServe(bindAddress, proxyMux)
+	var nodeIPs []net.IP
+	for _, family := range []v1.IPFamily{v1.IPv4Protocol, v1.IPv6Protocol} {
+		npa := proxyutil.NewNodePortAddresses(family, cidrStrings)
+		ips, err := npa.GetNodeIPs(proxyutil.RealNetwork{})
```
Using `proxyutil.NodePortAddresses` is wrong here; that has some semantics that don't necessarily apply here, and might have other nodeport-specific behavior in the future.

Possibly some helper should be split out and shared by NodePortAddresses and this code.

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -610,10 +600,10 @@ func (s *ProxyServer) birthCry() {
 //
 // The order of precedence is:
-//  1. if bindAddress is not 0.0.0.0 or ::, then it is used as the primary IP.
+//  1. if nodeIPOverride is not 0.0.0.0 or ::, then it is used as the primary IP.
```
The docs need to be updated more than just that

danwinship reviewed (commented):
blah, we need to split this up some more to make it more reviewable.

You could start reorganizing `pkg/proxy/apis/config/types.go` before introducing v1alpha2. Like, one commit to introduce the `Linux` section and move stuff into there, one to introduce the shared/merged syncPeriod/minSyncPeriod, etc. Then each piece is smaller and easier to review.

aroradaman commented:
/test pull-kubernetes-unit 
```
k8s.io/kubectl/pkg/cmd: logs TestLog/stateful_set_logs_with_all_pods
```

aroradaman committed:
pkg/proxy: refactor npa.GetNodeIPs

Signed-off-by: Daman Arora <aroradaman@gmail.com>

neolit123 reviewed (commented):
@aroradaman @danwinship something that i did not see covered in the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/784-kube-proxy-component-config) is what is the expected deprecation and removal path for existing alphas, and more specifically v1alpha1, which has been in production for almost 10 years.

i don't think the guidance here is applicable:
https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api

my recommendation is to:
- mark v1alpha1 as deprecated with the release of v1alpha2
- keep v1alpha1 for a minimum of 3 releases
- remove v1alpha1
- additional alphas/betas can be added async

aroradaman commented:
> @aroradaman @danwinship something that i did not see covered in the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/784-kube-proxy-component-config) is what is the expected deprecation and removal path for existing alphas, and more specifically v1alpha1, which has been in production for almost 10 years.
> 
> i don't think the guidance here is applicable: https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api
> 
> my recommendation is to:
> 
> * mark v1alpha1 as deprecated with the release of v1alpha2
> * keep v1alpha1 for a minimum of 3 releases
> * remove v1alpha1
> * additional alphas/betas can be added async

@neolit123 https://github.com/kubernetes/enhancements/pull/4337#discussion_r1460863776

neolit123 commented:
> > @aroradaman @danwinship something that i did not see covered in the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/784-kube-proxy-component-config) is what is the expected deprecation and removal path for existing alphas, and more specifically v1alpha1, which has been in production for almost 10 years.
> > i don't think the guidance here is applicable: https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api
> > my recommendation is to:
> > 
> > * mark v1alpha1 as deprecated with the release of v1alpha2
> > * keep v1alpha1 for a minimum of 3 releases
> > * remove v1alpha1
> > * additional alphas/betas can be added async
> 
> @neolit123 [kubernetes/enhancements#4337 (comment)](https://github.com/kubernetes/enhancements/pull/4337#discussion_r1460863776)

thanks for linking the discussion,

while we have to be cautious about removing v1alpha1 and breaking users, i don't think an alpha component config API should be maintained indefinitely. given it existed for so long it can be treated as GA and be removed in e.g. 4 or more releases.

the ecosystem still needs a forcing function to migrate to the eventual latest stable (e.g. v1).

danwinship commented:
oh, wow, I'd been avoiding this because it was so hard to review before, but with the new refactoring this is great.

So how about you split off the first three commits (splitting out options.go) into their own PR, and that should be ready to merge right away. And then meanwhile fix the rebase conflict on the rest of it and repush that and I can re-review. I just skimmed through everything now and I'm pretty sure it's all good, though I'm going to have to carefully think about all the conversion stuff for at least one of them... (Maybe split out the Linux and Windows sections part into their own PR as well, and then we can make sure everything is right there, and then the rest of this PR won't need as much review.)

aroradaman committed:
kube-proxy: internal config: consolidate SyncPeriod and MinSyncPeriod

Consolidate SyncPeriod and MinSyncPeriod for internal configuration
of kube-proxy adhering to the v1alpha2 version specifications as
detailed in https://kep.k8s.io/784.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman commented on pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/roundtrip/default/v1alpha1.yaml:
```diff
@@ -54,8 +54,8 @@ mode: ""
   masqueradeBit: 14
-  minSyncPeriod: 1s
-  syncPeriod: 30s
+  minSyncPeriod: 0s
```
@danwinship In conversion from internal -> v1alpha1 I'm currently setting `syncPeriod` and `minSyncPeriod` based on proxy-mode. Roundtrip will set these to zero values for other backends which I guess makes sense. 

Currently, we don't have a default value for `ipvs.minSyncPeriod` so the change to zero value is not visible (already zero) in this commit.

This would have been cleaner if 'omitempty' had been set for the v1alpha1 config.

aroradaman committed:
kube-proxy: internal config: refactor ClusterCIDR

Refactor ClusterCIDR for internal configuration of kube-proxy
adhering to the v1alpha2 version specifications as detailed in
https://kep.k8s.io/784.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
kube-proxy: internal config: remove PortRange

Remove PortRange for internal configuration of kube-proxy
adhering to the v1alpha2 version specifications as detailed in
https://kep.k8s.io/784.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
kube-proxy: internal config: add IPVS.MasqueradeBit

Add IPVS.MasqueradeBit for internal configuration of kube-proxy
adhering to the v1alpha2 version specifications as detailed in
https://kep.k8s.io/784.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
kube-proxy: internal config: refactor BindAddress

Replace BindAddress with NodeIPOverride for internal configuration of
kube-proxy adhering to the v1alpha2 version specifications as detailed
in https://kep.k8s.io/784.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
kube-proxy: use netutils.MultiListen for healthz and metrics server

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
kube-proxy: internal config: refactor Healthz and Metrics Address

Refactor Healthz with Metrics Address for internal configuration of
kube-proxy adhering to the v1alpha2 version specifications as detailed
in https://kep.k8s.io/784.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman commented:
`pull-kubernetes-e2e-kind-ipv6` fails at binding healthz.

```
"Healthz server failed" err="failed to start proxier healthz on [[::1]:10256 [fc00:f853:ccd:e793::2]:10256 [fe80::42:acff:fe12:2]:10256]: listen tcp [fe80::42:acff:fe12:2]:10256: bind: invalid argument"
```

danwinship commented on cmd/kube-proxy/app/server_test.go:
```diff
@@ -142,71 +142,71 @@ func Test_detectNodeIPs(t *testing.T) {
 	}{
 		{
 			name:           "Bind address IPv4 unicast address and no Node object",
 			rawNodeIPs:     nil,
-			bindAddress:    "10.0.0.1",
+			nodeIPOverride: []string{"10.0.0.1"},
```
should probably rename the tests to not say "Bind address" too

danwinship commented on pkg/proxy/apis/config/fuzzer/fuzzer.go:
```diff
@@ -33,7 +33,7 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 		func(obj *kubeproxyconfig.KubeProxyConfiguration, c fuzz.Continue) {
 			c.FuzzNoCustom(obj)
-			obj.BindAddress = fmt.Sprintf("%d.%d.%d.%d", c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(256))
+			obj.NodeIPOverride = []string{fmt.Sprintf("%d.%d.%d.%d", c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(256))}
```
should possibly fuzz to 0-2 values...
(@aojea did we end up adding an ip-address-fuzzing util somewhere? I feel like we talked about that as part of the ServiceCIDR stuff)

aroradaman replied:
BindAddress is not meant to be comma-separated, If I fuzz between 0-2 then I will lose data in roundtrip conversion tests.

danwinship replied:
Yeah... I think we want to just give up on the idea that you can roundtrip v1alpha1 with a fuzzing step in the middle. We'll want that to work for v1alpha2, but the goal here is to explicitly freeze-and-abandon the v1alpha1 API.

(I'm not sure what this means for the unit tests...)

aroradaman commented on pkg/proxy/apis/config/fuzzer/fuzzer.go:
```diff
@@ -44,6 +44,7 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 			obj.IPTables.MasqueradeBit = ptr.To(c.Int31())
 			obj.IPTables.LocalhostNodePorts = ptr.To(c.RandBool())
 			obj.NFTables.MasqueradeBit = ptr.To(c.Int31())
+			obj.IPVS.MasqueradeBit = obj.IPTables.MasqueradeBit
```
yup.

aroradaman commented on cmd/kube-proxy/app/options.go:
```diff
@@ -110,7 +111,7 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 
 	fs.StringVar(&o.hostnameOverride, "hostname-override", o.hostnameOverride, "If non-empty, will be used as the name of the Node that kube-proxy is running on. If unset, the node name is assumed to be the same as the node's hostname.")
-	fs.Var(&utilflag.IPVar{Val: &o.config.BindAddress}, "bind-address", "Overrides kube-proxy's idea of what its node's primary IP is. Note that the name is a historical artifact, and kube-proxy does not actually bind any sockets to this IP. This parameter is ignored if a config file is specified by --config.")
+	fs.Var(&utilflag.IPVar{Val: &o.bindAddress}, "bind-address", "Overrides kube-proxy's idea of what its node's primary IP is. Note that the name is a historical artifact, and kube-proxy does not actually bind any sockets to this IP. This parameter is ignored if a config file is specified by --config.")
```
Yes, we planned to maintain `--node-ip-override` and `--hostname-override` for v1alpha2 https://github.com/kubernetes/kubernetes/pull/121830#discussion_r1509125848

`bind-address` is only for backward compatability.

dims commented:
still need this @aroradaman ? please rebase if so

dims commented:
/close
